services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py

```diff
@@ -95,16 +99,43 @@ def _inject_paths_mappings(
         service_content["environment"] = _environment_section.export_as_list(env_vars)
 
 
-def _replace_env_vars_in_compose_spec(
-    stringified_service_spec: str, resolved_registry_url: str, service_tag: str
-) -> str:
-    stringified_service_spec = stringified_service_spec.replace(
-        MATCH_SIMCORE_REGISTRY, resolved_registry_url
-    )
-    stringified_service_spec = stringified_service_spec.replace(
-        MATCH_SERVICE_VERSION, service_tag
-    )
-    return stringified_service_spec
+def _update_resource_limits_and_reservations(
+    service_resources: ServiceResources, service_spec: ComposeSpecLabel
+) -> None:
+    # example: '2.3' -> 2 ; '3.7' -> 3
+    docker_compose_major_version: int = int(service_spec["version"].split(".")[0])
```

Comment:
THOUGHT: ``ComposeSpecLabel`` was in a "first round" defined as a ``Dict[str, Any]``.  That is ok, because it was the first time we use it ... but as soon as it becomes more and more used, we should add more and more constraints to it. For instance, based on the usage here, I would extend  in``models_library/service_settings_labels.py``
```python

class ComposeSpecLabel(TypedDict):
    version: str
    services: Dict[str, Any]

```
then we collaboratively do the code safer.

Same applies for any ``dict`` that holds **structured data** , i.e. they all should slowly be defined as ``TypedDict``s

Reply:
any input on thisthought?

Reply:
I think that once we move the generated docker-compose spec pydantic models to the models library we can attempt to refactor this.